driver-core/src/main/java/com/datastax/driver/core/Message.java

```diff
@@ -170,21 +171,25 @@ public UUID getTracingId() {
         }
     }
 
-    public static class ProtocolDecoder extends OneToOneDecoder {
+    @ChannelHandler.Sharable
+    public static class ProtocolDecoder extends MessageToMessageDecoder<Frame> {
 
-        public Object decode(ChannelHandlerContext ctx, Channel channel, Object msg) {
-            assert msg instanceof Frame : "Expecting frame, got " + msg;
-
-            Frame frame = (Frame)msg;
+        @Override
+        protected void decode(ChannelHandlerContext ctx, Frame frame, List<Object> out) throws Exception {
             boolean isTracing = frame.header.flags.contains(Frame.Header.Flag.TRACING);
             UUID tracingId = isTracing ? CBUtil.readUUID(frame.body) : null;
 
             Response response = Response.Type.fromOpcode(frame.header.opcode).decoder(frame.header.version).decode(frame.body);
-            return response.setTracingId(tracingId).setStreamId(frame.header.streamId);
+            ReferenceCountUtil.release(frame.body);
```

Comment:
just to make sure I understand this, this just release the frame body earlier since we're done with it at this point now that we got a response from it?